### PR TITLE
feat(dashboard): Dashboards backend implementation

### PR DESCRIPTION
**Pull request by keremsahin1**
Manually done the integration tests after running dev Dockers. 

1. Ingested sample data for both chart and dashboard entities through ingest endpoint.
2. Checked ingested data through get and get_all endpoints.
3. Tested search & autocomplete endpoints.
4. Checked Neo4j graph to see if nodes & edges are generated as expected.

Also tested and validated the MCE ingestion pipeline by emitting sample MCEs using `mce_cli`.

## Checklist
- [X] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [X] Tests for the changes have been added/updated (if applicable)
- [X] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: gms/impl/src/main/java/com/linkedin/metadata/resources/dashboard/Charts.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+@RestLiCollection(name = "charts", namespace = "com.linkedin.chart", keyName = "key")
+public class Charts extends BaseSearchableEntityResource<
+    // @formatter:off
+    ChartKey,
+    Chart,
+    ChartUrn,
+    ChartSnapshot,
+    ChartAspect,
+    ChartDocument> {
+  // @formatter:on
```
Comment (jplaisted) [inline gms/impl/src/main/java/com/linkedin/metadata/resources/dashboard/Charts.java:56]: Nit: align with formatter:off?

File: gms/impl/src/main/java/com/linkedin/metadata/resources/dashboard/Dashboards.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+@RestLiCollection(name = "dashboards", namespace = "com.linkedin.dashboard", keyName = "key")
+public class Dashboards extends BaseSearchableEntityResource<
+    // @formatter:off
+    DashboardKey,
+    Dashboard,
+    DashboardUrn,
+    DashboardSnapshot,
+    DashboardAspect,
+    DashboardDocument> {
+  // @formatter:on
```
Comment (jplaisted) [inline gms/impl/src/main/java/com/linkedin/metadata/resources/dashboard/Dashboards.java:55]: ditto

File: li-utils/src/main/java/com/linkedin/common/urn/ChartUrn.java [hunk 3]
```diff
@@ -31,9 +31,10 @@ public String getChartIdEntity() {
   }
 
   public static ChartUrn createFromString(String rawUrn) throws URISyntaxException {
-    Urn urn = new Urn(rawUrn);
+    final Urn urn = new Urn(rawUrn);
     validateUrn(urn, ENTITY_TYPE);
-    String[] urnParts = urn.getContent().split(",");
+    final String content = urn.getContent();
+    final String[] urnParts = content.substring(1, content.length() - 1).split(",");
```
Comment (jplaisted) [inline li-utils/src/main/java/com/linkedin/common/urn/ChartUrn.java:37]: What changed here and why?

File: metadata-models/src/main/pegasus/com/linkedin/metadata/search/DashboardDocument.pdl [hunk 4]
```diff
@@ -41,5 +36,5 @@ record DashboardDocument includes BaseDocument {
   /**
    * Access level for the dashboard
    */
-  access: optional string
+  access: optional AccessLevel
```
Comment (jplaisted) [inline metadata-models/src/main/pegasus/com/linkedin/metadata/search/DashboardDocument.pdl:39]: Can we leave model changes to their own PRs, outside of implementation? Also, we just had a round of design reviews for this, why did this change (blanket question that applies to other changes like removing chartIds)

**Review by jplaisted — CHANGES_REQUESTED** [review]

File: li-utils/src/main/java/com/linkedin/common/urn/ChartUrn.java (hunk 3 above)
Comment (keremsahin1) [reply li-utils/src/main/java/com/linkedin/common/urn/ChartUrn.java:37]: This is a bug fix. I realized that parsing logic is not correct. Content contains opening and closing parenthesis. We need to get rid of them.

File: metadata-models/src/main/pegasus/com/linkedin/metadata/search/DashboardDocument.pdl (hunk 4 above)
Comment (keremsahin1) [reply metadata-models/src/main/pegasus/com/linkedin/metadata/search/DashboardDocument.pdl:39]: This is actually an innocent model change :) I earlier forgot that we support enums in the document models and set the types for a couple of enum fields to string. Now, setting them back to their actual types for better type checking.

Comment (keremsahin1) [reply metadata-models/src/main/pegasus/com/linkedin/metadata/search/DashboardDocument.pdl:39]: Regarding removing chartIds: I originally thought that we can store dashboard <-> chart <-> dataset (or metric) relationships in the search index to support queries like "what are the downstream dashboards of this dataset?". However, any kind of relationship query could be supported by graph (that's what it's for in the end) and think that no need to store these relations in the search index. WDYT?

**Comment by mars-lan** [comment]
@grantatspothero you may be interested in reviewing this.

File: metadata-models/src/main/pegasus/com/linkedin/metadata/search/DashboardDocument.pdl (hunk 4 above)
Comment (grantatspothero) [reply metadata-models/src/main/pegasus/com/linkedin/metadata/search/DashboardDocument.pdl:39]: +1 on leaving the modeling changes to separate PRs. I am not an elasticsearch expert, but I can leave feedback on models and how they would impact our usage. Right now the PR is pretty large and it makes that more difficult.

File: gms/api/src/main/pegasus/com/linkedin/dashboard/ChartKey.pdl [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  @validate.strlen = {
+    "max" : 20,
+    "min" : 1
+  }
+  tool: string
+
+  /**
+   * Unique id of the chart assigned by the specific dashboard tool
+   */
+  @validate.strlen = {
```
Comment (grantatspothero) [inline gms/api/src/main/pegasus/com/linkedin/dashboard/ChartKey.pdl:19]: SpotHero uses Looker. Looker has different types of entities. For example:
- "Looks", which are just single visualizations.
- "Dashboards", which are collections of visualizations displayed on one page.
- "Boards", which are collections of Looks or Dashboards.

Each has a different URL structure:
- Looks: `${BASE_URI}/looks/${LOOK_ID}`
- Dashboards: `${BASE_URI}/dashboards/${DASHBOARD_ID}`
- Boards: `${BASE_URI}/browse/boards/${BOARD_ID}`

I can forsee some problems with `chartId` and trying to distinguish between look IDs vs dashboard IDs vs board IDs. There are different ways to resolve this problem, but the simplest would be getting rid of the idea of a `chartId` and instead have a `chartUri` and just storing the URI of the visualization tool where the component exists.

File: metadata-models/src/main/pegasus/com/linkedin/metadata/search/DashboardDocument.pdl (hunk 4 above)
Comment (jplaisted) [reply metadata-models/src/main/pegasus/com/linkedin/metadata/search/DashboardDocument.pdl:39]: What does "innocent" mean? Unrelated? More of a reason to put in a different PR, imo.

File: gms/api/src/main/snapshot/com.linkedin.dashboard.dashboards.snapshot.json [hunk 6]
```diff
@@ -0,0 +1,600 @@
...
+          "type" : "string",
+          "doc" : "Title of the dashboard"
+        }, {
+          "name" : "description",
+          "type" : "string",
+          "doc" : "Detailed description about the dashboard"
+        }, {
+          "name" : "charts",
+          "type" : {
+            "type" : "array",
```
Comment (grantatspothero) [inline gms/api/src/main/snapshot/com.linkedin.dashboard.dashboards.snapshot.json:264]: How does Looker fit into the "a dashboard has multiple charts" modeling. 

Looker lets you do a couple of different things:
- Add a look to a dashboard as a tile.
- Add an explore (with certain filters applied) to a dashboard as a tile. 

Is the idea that an "explore" with certain filters applied would be modeled as a "chart" and a "look" would be modeled as a chart? As I mentioned above, there can be some funkiness because Looker treats explores and looks differently.

File: metadata-models/src/main/pegasus/com/linkedin/metadata/search/DashboardDocument.pdl (hunk 4 above)
Comment (keremsahin1) [reply metadata-models/src/main/pegasus/com/linkedin/metadata/search/DashboardDocument.pdl:39]: Sure, let me send model changes as a separate PR ;)

File: gms/api/src/main/pegasus/com/linkedin/dashboard/ChartKey.pdl (hunk 5 above)
Comment (keremsahin1) [reply gms/api/src/main/pegasus/com/linkedin/dashboard/ChartKey.pdl:19]: Thanks for your review @grantatspothero and getting a review from a Looker expert is definitely very helpful (I never used Looker) :)

In the design phase, I identified two entity types - [charts & dashboards](https://github.com/linkedin/datahub/tree/master/docs/rfc/active/1778-dashboards#entities), which are common between different dashboarding tools. Please also review RFC for dashboards.

From your comment, I understand that `board` would also be considered as a separate entity type (at least for Looker). Is my understanding correct? We can add that as an entity but do you think it's a generic concept that can be applied to any other dashboarding tool or is it Looker specific?

File: gms/api/src/main/snapshot/com.linkedin.dashboard.dashboards.snapshot.json (hunk 6 above)
Comment (keremsahin1) [reply gms/api/src/main/snapshot/com.linkedin.dashboard.dashboards.snapshot.json:264]: Again, appreciate the expert review. Probably I wasn't quite careful in my study to explore "explore"s in Looker :) Do you think it should be a separate entity? Or, is it basically another look with certain filters applied?

File: metadata-models/src/main/pegasus/com/linkedin/metadata/search/DashboardDocument.pdl (hunk 4 above)
Comment (keremsahin1) [reply metadata-models/src/main/pegasus/com/linkedin/metadata/search/DashboardDocument.pdl:39]: New PR for model updates only: https://github.com/linkedin/datahub/pull/1932

File: gms/factories/src/main/java/com/linkedin/dashboard/factory/ChartDaoFactory.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
+package com.linkedin.dashboard.factory;
```
Comment (mars-lan) [inline gms/factories/src/main/java/com/linkedin/dashboard/factory/ChartDaoFactory.java:1]: Please rename namespace due to https://github.com/linkedin/datahub/pull/1916

File: gms/factories/src/main/java/com/linkedin/dashboard/factory/ChartSearchDaoFactory.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
+package com.linkedin.dashboard.factory;
```
Comment (mars-lan) [inline gms/factories/src/main/java/com/linkedin/dashboard/factory/ChartSearchDaoFactory.java:1]: ditto

File: gms/factories/src/main/java/com/linkedin/dashboard/factory/DashboardDaoFactory.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
+package com.linkedin.dashboard.factory;
```
Comment (mars-lan) [inline gms/factories/src/main/java/com/linkedin/dashboard/factory/DashboardDaoFactory.java:1]: ditto

File: gms/factories/src/main/java/com/linkedin/dashboard/factory/DashboardSearchDaoFactory.java [hunk 10]
```diff
@@ -0,0 +1,28 @@
+package com.linkedin.dashboard.factory;
```
Comment (mars-lan) [inline gms/factories/src/main/java/com/linkedin/dashboard/factory/DashboardSearchDaoFactory.java:1]: ditto

**Review by mars-lan — CHANGES_REQUESTED** [review]

File: gms/api/src/main/pegasus/com/linkedin/dashboard/ChartKey.pdl (hunk 5 above)
Comment (grantatspothero) [reply gms/api/src/main/pegasus/com/linkedin/dashboard/ChartKey.pdl:19]: `Board` is a very Looker-specific concept and totally makes sense to leave out as a first pass 👍

Comment (keremsahin1) [reply gms/api/src/main/pegasus/com/linkedin/dashboard/ChartKey.pdl:19]: > it might make sense to store a chartURI explicitly
How is `chartURI` defined? Is it including the `BASE_URI` in your original definition? URN should have two main characteristics: unique & not changing. I feel like `BASE_URI` might be changing like v1 becomes v2 etc or your looker server's name (or IP) changes. If that happens, you'll have two copies of metadata for the same chart/dashboard.

> constructing a URI from a chartid is non-trivial based on the type of chart
This is where I'm confused at. If we store `LOOK_ID` as chartId in the URN, are you not able construct a chart URI using ${BASE_URI}/looks/${chartId}. What do you mean by chart types? We have a separate URN for dashboards. And, we will use it for Looker dashboards. Only thing we're missing is `boards` at this point.

Comment (grantatspothero) [reply gms/api/src/main/pegasus/com/linkedin/dashboard/ChartKey.pdl:19]: > How is chartURI defined? Is it including the BASE_URI in your original definition? URN should have two main characteristics: unique & not changing. I feel like BASE_URI might be changing like v1 becomes v2 etc or your looker server's name (or IP) changes. If that happens, you'll have two copies of metadata for the same chart/dashboard. I don't think we want that?

Yes I was including the base uri in the chart uri. If datahub does not explicitly store a chartURI then how do you map from a unique chart id back to a URI? It seems important for datahub to surface a link to the chart since a chart id of eg: `522` does not actually mean anything. 

If datahub does not store a chartURI, then it would be on datahub to understand that a `chart` in Looker actually means a `Look` and then template in the chart id into some looker look URI base template. See my below comment for why this can get complicated when considering "looks" vs "explores"

File: gms/api/src/main/snapshot/com.linkedin.dashboard.dashboards.snapshot.json (hunk 6 above)
Comment (grantatspothero) [reply gms/api/src/main/snapshot/com.linkedin.dashboard.dashboards.snapshot.json:264]: "Explores" are kind of like views in Looker, they represent virtual tables with pre-joined relationships between the dependent tables so people can click around the Looker UI filters to generate queries. 
https://docs.looker.com/reference/explore-params/explore

As I mentioned above, Looker lets you add charts to a dashboard in two different ways:
- Add a look to a dashboard. 
- Add an explore with certain filters applied to a dashboard. This is essentially like a "Look" but it is unnamed (while a Look is a named entity). It also has a different URI structure. 

Look URI:
` ${BASE_URI}/looks/${LOOK_ID}`

Explore URI:
`${BASE_URI}/explore/${explore_name}/history?qid=${unique_query_id_hash}`

Comment (grantatspothero) [reply gms/api/src/main/snapshot/com.linkedin.dashboard.dashboards.snapshot.json:264]: Also if it helps, here is a public Looker instance you can play around with that contains covid-19 data. Might help to understand the Looker concepts:
https://covid19response.cloud.looker.com/embed/dashboards-next/51

File: gms/api/src/main/pegasus/com/linkedin/dashboard/ChartKey.pdl (hunk 5 above)
Comment (keremsahin1) [reply gms/api/src/main/pegasus/com/linkedin/dashboard/ChartKey.pdl:19]: We have couple use cases of external links in internal DataHub. The approach we've taken for that is to basically store `BASE_URI` as a global config to generate the external links. But this has the underlying assumption that combination of BASE_URI and stored chart id will be enough to uniquely produce the link for the chart. If not, that means we need to store more metadata for the chart.

After reading your below comment, I started to think of looks & explores as two different customizations of chart (think of chart as the parent class, looks & explores as child classes). But, my understanding could still be limited.

File: gms/api/src/main/snapshot/com.linkedin.dashboard.dashboards.snapshot.json (hunk 6 above)
Comment (keremsahin1) [reply gms/api/src/main/snapshot/com.linkedin.dashboard.dashboards.snapshot.json:264]: Thanks for the pointers. I started to think that looks and explores are two different customizations of "chart". Your thinking is right, to be able to differentiate between looks vs explores, part of URI could be put into the URN definition. For example: chartId could be `looks/${LOOK_ID}` for looks and `explore/${explore_name}` for explores.

On the other hand, there is another way to approach this problem. Modelling dashboard metadata was a little tricky because we thought that the models provided should be easily applicable/extensible to any dashboarding tool out there. So, we came up with two entities which are dashboard and chart. These are most basic building blocks for a dashboarding platform. If Looker has an additional entity (I guess the concept of `board` could be good example), extra modelling will be needed for that. Or if Looker has two very different types of charts, extra modelling *might* be needed like `TypeAChart extends Chart` and `TypeBChart extends Chart`. One thing to be careful here, Pegasus doesn't quite have inheritance but it has composition: https://linkedin.github.io/rest.li/pdl_schema#including-fields

File: gms/api/src/main/pegasus/com/linkedin/dashboard/ChartKey.pdl (hunk 5 above)
Comment (grantatspothero) [reply gms/api/src/main/pegasus/com/linkedin/dashboard/ChartKey.pdl:19]: I think different chart types with customizable base URIs makes sense. Will allow support different dashboarding tools.

Another consideration is supporting multiple dashboard instances per datahub install. Is this something you are trying to support? For example, SpotHero uses two Looker instances (one large instance for internal use, and one small instance for external use) and the base URIs would need to be different even though both are "Looker" charts.

Comment (keremsahin1) [reply gms/api/src/main/pegasus/com/linkedin/dashboard/ChartKey.pdl:19]: According to our meeting, I'll change the definition of `chartId` as `chartURI`.

**Review by shirshanka — APPROVED** [review]
LGTM!
